packages/loopring_v3/contracts/converters/BaseConverter.sol

```diff
@@ -0,0 +1,202 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2017 Loopring Technology Limited.
+pragma solidity ^0.7.0;
+
+import "../core/iface/IExchangeV3.sol";
+import "../lib/Claimable.sol";
+import "../lib/Drainable.sol";
+import "../lib/ERC20.sol";
+import "../lib/AddressUtil.sol";
+import "../lib/ERC20SafeTransfer.sol";
+import "../lib/MathUint.sol";
+import "../lib/LPERC20.sol";
+
+
+/// @author Brecht Devos - <brecht@loopring.org>
+abstract contract BaseConverter is LPERC20, Claimable, Drainable
+{
+    using AddressUtil       for address;
+    using ERC20SafeTransfer for address;
+    using MathUint          for uint;
+
+    event ConversionSuccess (uint amountIn, uint amountOut);
+    event ConversionFailed  (string reason);
+
+    IExchangeV3        public  immutable exchange;
+    IDepositContract   public  immutable depositContract;
+
+    bool               public  initialized;
+
+    address            public  tokenIn;
+    address            public  tokenOut;
+
+    bool               public  failed;
+
+    modifier onlyFromExchangeOwner()
+    {
+        require(msg.sender == exchange.owner(), "UNAUTHORIZED");
+        _;
+    }
+
+    constructor(
+        IExchangeV3 _exchange
+        )
+    {
+        exchange = _exchange;
+        depositContract = _exchange.getDepositContract();
+    }
+
+    function initialize(
+        string memory _name,
+        string memory _symbol,
+        uint8         _decimals,
+        address       _tokenIn,
+        address       _tokenOut
+        )
+        external
+    {
+        require(!initialized, "ALREADY_INITIALIZED");
+        initializeToken(_name, _symbol, _decimals);
+
+        tokenIn = _tokenIn;
+        tokenOut = _tokenOut;
+
+        initialized = true;
+    }
+
+    function deposit(
+        uint96          amountIn,
+        uint96          minAmountOut,
+        bytes  calldata customData
+        )
+        external
+        payable
+        onlyFromExchangeOwner
+    {
+        require(totalSupply == 0);
+
+        // Converter specific logic, which can fail
+        try BaseConverter(this).convertExternal(amountIn, minAmountOut, customData)
+            returns (uint amountOut) {
+            failed = false;
+            emit ConversionSuccess(amountIn, amountOut);
+        } catch Error(string memory reason) {
+            failed = true;
+            emit ConversionFailed(reason);
+        } catch {
+            failed = true;
+            emit ConversionFailed("unknown");
+        }
+
+        // Mint pool tokens representing each user's share in the pool, with 1:1 ratio
+        _mint(address(this), amountIn);
+
+        // Repay the flash mint used to give user's their share on L2
+        _repay(address(this), amountIn);
```

Comment:
The new-minted pool tokens will be eventually transferred to the exchange's deposit contract, correct?

Reply:
Not sure if I understand the question correctly, but yes that's what the repay function will do. They are already on L2, the repay function makes sure the tokens actually exist on L1.